Create article: Return Razor component HTML from a web API using RazorComponentResult (#37405)
dotnet/AspNetCore.Docs: aspnetcore/blazor/components/return-component-html-from-a-web-api.md

---
title: Return Razor component HTML from a web API in ASP.NET Core Blazor
ai-usage: ai-assisted
author: guardrex
description: Learn how to return HTML (as a string) from a web API (Minimal API) using the RazorComponentResult class.

> monikerRange: '>= aspnetcore-6.0'

Reviewer: I think the `RazorComponentResult` was introduced in ASP.NET Core 8.0 [API Doc](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.httpresults.razorcomponentresult?view=aspnetcore-10.0)

Reviewer 2: Thanks, @bassem-mf! Correct ... and I wish Copilot was smart enough to have figured that out. Looks like us poor humans are still needed around here! 😆 

@copilot ... Change this line to make the article apply to .NET 8 or later (`monikerRange: '>= aspnetcore-8.0'`).

> monikerRange: '>= aspnetcore-6.0'

Reviewer 2: ```suggestion
monikerRange: '>= aspnetcore-8.0'
```

ms.author: wpickett
ms.date: 07/31/2026
uid: blazor/components/return-component-html-from-a-web-api
---
# Return Razor component HTML from a web API in ASP.NET Core Blazor

[!INCLUDE[](~/includes/not-latest-version.md)]

> 

Reviewer 2: @copilot ... Remove Lines 13 and 14.

```suggestion
```

---
title: Return Razor component HTML from a web API in ASP.NET Core Blazor
ai-usage: ai-assisted
author: guardrex
description: Learn how to return HTML (as a string) from a web API (Minimal API) using the RazorComponentResult class.
monikerRange: '>= aspnetcore-6.0'
ms.author: wpickett
ms.date: 07/31/2026
uid: blazor/components/return-component-html-from-a-web-api
---
# Return Razor component HTML from a web API in ASP.NET Core Blazor

[!INCLUDE[](~/includes/not-latest-version.md)]

> Components returned via <xref:Microsoft.AspNetCore.Http.HttpResults.RazorComponentResult`1> are rendered as static HTML strings. Features requiring an active SignalR connection for component interactivity can't execute in this context.

Reviewer 2: ```suggestion
This article demonstrates how to return Razor component HTML (as a `string`) from a web API (Minimal API) using the <xref:Microsoft.AspNetCore.Http.HttpResults.RazorComponentResult%601> class.

Components returned via <xref:Microsoft.AspNetCore.Http.HttpResults.RazorComponentResult%601> are rendered as static HTML strings. Features requiring an active SignalR connection for component interactivity can't execute in this context.
```

## Place the Razor components into the web API

`Components/Greeting.razor`:

```razor
<h2>Greeting component</h2>

<p>Hello, world!</p>
```

`Components/UserData.razor`:

```razor
<h2>UserData component</h2>

<p>Viewing details for User ID: @UserId</p>

@code {
    [Parameter]
    public int UserId { get; set; }
}
```

## Add Razor component services and endpoints to the web API

> For components with parameters, pass dictionary keys that exactly match the names of properties decorated with `[Parameter]` on the component.

Reviewer 2: ```suggestion
For components with parameters, pass dictionary keys that exactly match the names of properties decorated with the [`[Parameter]` attribute](xref:Microsoft.AspNetCore.Components.ParameterAttribute).
```

---
title: Return Razor component HTML from a web API in ASP.NET Core Blazor
ai-usage: ai-assisted
author: guardrex
description: Learn how to return HTML (as a string) from a web API (Minimal API) using the RazorComponentResult class.
monikerRange: '>= aspnetcore-8.0'
ms.author: wpickett
ms.date: 07/31/2026
uid: blazor/components/return-component-html-from-a-web-api
---
# Return Razor component HTML from a web API in ASP.NET Core Blazor

This article demonstrates how to return Razor component HTML (as a `string`) from a web API (Minimal API) using the <xref:Microsoft.AspNetCore.Http.HttpResults.RazorComponentResult> class.

Components returned via <xref:Microsoft.AspNetCore.Http.HttpResults.RazorComponentResult> are rendered as static HTML strings. Features requiring an active SignalR connection for component interactivity can't execute in this context.

> ## Place the Razor components into the web API

Reviewer 3: ```suggestion
## Add the Razor components to the web API
```
Very minor.  Seemed less awkward this way.

`Components/Greeting.razor`:

```razor
<h2>Greeting component</h2>

<p>Hello, world!</p>
```

`Components/UserData.razor`:

```razor
<h2>UserData component</h2>

<p>Viewing details for User ID: @UserId</p>

@code {
    [Parameter]
    public int UserId { get; set; }
}
```

## Add Razor component services and endpoints to the web API

For components with parameters, pass dictionary keys that exactly match the names of properties decorated with the [`[Parameter]` attribute](xref:Microsoft.AspNetCore.Components.ParameterAttribute).

```csharp
using Microsoft.AspNetCore.Http.HttpResults;
using MinimalAPI.Components;

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddRazorComponents();

var app = builder.Build();

app.MapGet("/hello", () => new RazorComponentResult<Greeting>());

app.MapGet("/user/{id:int}", (int id) =>
{
    var parameters = new Dictionary<string, object?>
    {
        { "UserId", id }
    };

    return new RazorComponentResult<UserData>(parameters);
});
```

## Consume the API and read the component HTML from the `RazorComponentResult`

> In the app's `Program` file, set up HTTP client services. The following example assumes that the web API responds to requests at `https://localhost:7286` using a [named `HttpClient`](xref:blazor/call-web-api#named-httpclient-with-ihttpclientfactory):

Reviewer 3: @guardrex: This "Consume the API" sectiion does not explicity state which Blazor hosting model the consuming component runs in, but the code behaves differently depending on the model. Should we clarify that?  Maybe in a prerequisite/note clarifying the target model?

```csharp
builder.Services.AddHttpClient("WebAPI", client =>
    client.BaseAddress = new Uri("https://localhost:7286"));
```

> In a component, inject <xref:System.Net.Http.IHttpClientFactory> and request the component HTML (as `string`s), casting each to a <xref:Microsoft.AspNetCore.Components.MarkupString> for rendering:

Reviewer 3: ```suggestion
In a component, inject an <xref:System.Net.Http.IHttpClientFactory> and request the component HTML (as `string`s), casting each to a <xref:Microsoft.AspNetCore.Components.MarkupString> for rendering:
```
Added missing "an".
